src/main/java/qtlog/FilesystemController/FileMonitor.java

```diff
@@ -26,16 +46,112 @@ public void notifyObserver() {
+            e.printStackTrace();
+        }
+        this.processEvents();
     }
 
-    //example code, delete later
-    public void setNewestFile(FileDTO newFile){
-        this.newestFile = newFile;
-        this.notifyObserver();
+    /**
+     * Register the given directory, and all its sub-directories, with the
+     * WatchService.
+     */
+    private void registerAll(final Path start) throws IOException {
+        // register directory and subdirectories
+        Files.walkFileTree(start, new SimpleFileVisitor<>() {
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+                throws IOException
+            {
+                registerKey(dir);
+                System.out.format("Registering dir: %s ...\n", dir);
+                return FileVisitResult.CONTINUE;
+            }
+        });
+    }
+
+    /**
+     * Register the given directory with the WatchService
+     */
+    private void registerKey(Path dir) throws IOException  {
+        WatchKey key = dir.register(watchservice, StandardWatchEventKinds.ENTRY_MODIFY);
+        if (trace) {
+            Path prev = keys.get(key);
+            if (prev == null) {
+                System.out.format("register: %s\n", dir);
+            } else {
+                if (!dir.equals(prev)) {
+                    System.out.format("update: %s -> %s\n", prev, dir);
+                }
+            }
+        }
+        keys.put(key, dir);
+    }
+
+    /**
+     * Creates a WatchService and registers the given directory
+     */
+    private void WatchDir(Path dir) throws IOException {
+        this.watchservice = FileSystems.getDefault().newWatchService();
+        this.keys = new HashMap<>();
+
+        System.out.format("Scanning %s ...\n", dir);
+        this.registerAll(dir);
+        System.out.println("Done.");
+
+        // enable trace after initial registration
+        this.trace = true;
+    }
+
+    /**
+     * Process all events for keys queued to the watcher
+     */
+    void processEvents() {
+        WatchEvent<Path> lastEvent = null;
+        while(true) {
+
+            // wait for key to be signalled
+            WatchKey key;
+            try {
+                key = watchservice.take();
+            } catch (InterruptedException x) {
+                return;
+            }
+
+            Path dir = keys.get(key);
+            if (dir == null) {
+                System.err.println("WatchKey not recognized!!");
+                continue;
+            }
+
+            for (WatchEvent<?> event: key.pollEvents()) {
+                
+                // Context for directory entry event is the file name of entry
+                WatchEvent<Path> ev = cast(event);
+                
+                Path name = ev.context();
+                Path child = dir.resolve(name);
+
+                if((lastEvent != StandardWatchEventKinds.ENTRY_MODIFY) && (name.toString().contains("evtc"))){
```

Comment:
This checks if the sequence of letters "evtc" is present in the name, but in theory it could be anywhere in the name not just the file extension. Instead use the .endsWith() method.
```java
if((lastEvent != StandardWatchEventKinds.ENTRY_MODIFY) && (name.endsWith(".evtc") || name.endsWith(".zevtc"))){
```

Reply:
Actually, it turns out neither of these work because right now none of the events contain the file extension at all. Plus we are getting dozens of modify events triggering as the file is being saved and by notifying upon the first one it's creating a race condition. I tried looking for a better approach but I've not found a solution yet.